File: java/src/main/structural/decorator/ConcreteDecoratorB.java
package structural.decorator;

// ConcreteDecoratorB: Adds different functionality to the component
public class ConcreteDecoratorB extends Decorator {

    public ConcreteDecoratorB(Component component) {
        super(component);
    }

    @Override
    public void operation() {
        super.operation();
        anotherAddedBehavior();
    }

    private void anotherAddedBehavior() {
        System.out.println("ConcreteDecoratorB: Adding another behavior.");
    }
}

// Example Usage:
// public class Main {
//     public static void main(String[] args) {
//         Component base = new ConcreteComponent();
//         
//         // Wrap the base component with decorators
//         Component decorated1 = new ConcreteDecoratorA(base);
//         Component decorated2 = new ConcreteDecoratorB(decorated1);
//
//         // Execute operation with all decorations applied
//         decorated2.operation();
//
//         // Expected Output:
//         // ConcreteComponent: Performing operation.
//         // ConcreteDecoratorA: Adding extra behavior.
//         // ConcreteDecoratorB: Adding another behavior.
//     }
// }
//
// Common Mistakes to Avoid:
//
// 1. Excessive Wrapping:
//    - Mistake: Applying too many decorators, making the call stack unnecessarily deep.
//    - Fix: Limit decoration to essential features.
//
// 2. Ignoring Base Class Responsibilities:
//    - Mistake: Not calling `super.operation()` in decorators, breaking expected behavior.
//    - Fix: Always call `super.operation()` unless explicitly overriding behavior.
//
// 3. Confusing Inheritance with Composition:
//    - Mistake: Using subclassing instead of decorators to add behavior.
//    - Fix: Favor composition (wrapping objects) over inheritance to keep code flexible.
//
// 4. Breaking Encapsulation:
//    - Mistake: Exposing concrete decorators in the client code, leading to dependency on specific implementations.
//    - Fix: Use the `Component` interface so that clients remain unaware of the specific decorators applied.

File: python/src/structural/decorator/concrete_decorator_b.py
from .decorator import Decorator


class ConcreteDecoratorB(Decorator):
    def operation(self) -> str:
        return f"ConcreteDecoratorB({super().operation()})"

# Example Usage:
# from concrete_component import ConcreteComponent
# from concrete_decorator_a import ConcreteDecoratorA
# from concrete_decorator_b import ConcreteDecoratorB
#
# component = ConcreteComponent()
# decorated = ConcreteDecoratorB(ConcreteDecoratorA(component))
# print(decorated.operation())
#
# # Expected Output:
# # ConcreteDecoratorB(ConcreteDecoratorA(ConcreteComponent))
#
# Common Mistakes to Avoid:
#
# 1. Skipping the Base Interface:
#    - Mistake: Not using a shared interface for components and decorators.
#    - Fix: Always derive decorators and components from a common abstract base
#      class.
#
# 2. Not Forwarding Calls:
#    - Mistake: Forgetting to delegate to the wrapped object.
#    - Fix: Always call `super().operation()` or equivalent inside your
#      decorators.
#
# 3. Overusing Decorators:
#    - Mistake: Nesting too many decorators can complicate debugging.
#    - Fix: Use layering when necessary, and document each decorator’s purpose.
#
# 4. Rigid Decorators:
#    - Mistake: Writing decorators that tightly couple to specific components.
#    - Fix: Use composition and interfaces to allow any component to be
#      decorated.
